File: rust_feedforward/src/loss.rs
#![allow(unused)]
use ndarray::{Array2, Axis};

pub fn cross_entropy_loss(predictions: &Array2<f32>, targets: &Array2<f32>) -> f32 {
    // small epsilon to prevent log(0)
    let epsilon = 1e-8;
    let clipped_predictions = predictions.mapv(|x| x.max(epsilon));
    let log_preds = clipped_predictions.mapv(|x| x.ln());
    let loss_matrix = targets * log_preds;
    let loss_per_example = loss_matrix.sum_axis(Axis(1)).mapv(|sum| -sum);
    loss_per_example.mean().unwrap()
}

// When using softmax followed by cross-entropy, this derivative
// simplifies to: predictions - targets.
pub fn cross_entropy_loss_derivative(
    predictions: &Array2<f32>,
    targets: &Array2<f32>,
) -> Array2<f32> {
    predictions - targets
}

File: python_feedforward/src/loss.py
import numpy as np

def cross_entropy_loss(predictions, targets, epsilon=1e-8):
    # prevent log(0)
    predictions_clipped = np.clip(predictions, epsilon, 1.0)
    loss_per_example = -np.sum(targets * np.log(predictions_clipped), axis=1)
    return np.mean(loss_per_example)

def cross_entropy_loss_derivative(predictions, targets):
    """
    For softmax combined with cross-entropy, the gradient simplifies to:
    predictions - targets.
    """
    return predictions - targets
